gnomad_qc/v3/create_release/make_var_annot_hists.py

```diff
@@ -1,61 +1,164 @@
-            minmax_dict[metric] = hl.struct(min=hl.agg.min(ht[metric]), max=hl.if_else(hl.agg.max(ht[metric])<1e10, hl.agg.max(ht[metric]), 1e10))
+        for metric in metrics:
+            minmax_dict[metric] = hl.struct(
+                min=hl.agg.min(ht.info[metric]),
+                max=hl.if_else(
+                    hl.agg.max(ht.info[metric]) < 1e10,
+                    hl.agg.max(ht.info[metric]),
+                    1e10,
+                ),
+            )
         minmax = ht.aggregate(hl.struct(**minmax_dict))
-        print(minmax)
+        logger.info(f"Metrics bounds: {minmax}")
     else:
-        # Aggregate hists over hand-tooled ranges
+        logger.info(
+            "Aggregating hists over ranges determined using first pass run..."
+        )
         hists = ht.aggregate(
             hl.array(
-                [hist_expr.annotate(metric=hist_metric) for hist_metric, hist_expr in hist_ranges_expr.items()]
+                [
+                    hist_expr.annotate(metric=hist_metric)
+                    for hist_metric, hist_expr in hist_ranges_expr.items()
+                ]
             ).extend(
                 hl.array(
                     hl.agg.group_by(
                         create_frequency_bins_expr(
-                            AC=ht.freq[1].AC,
-                            AF=ht.freq[1].AF
+                            AC=ht.freq[1].AC, AF=ht.freq[1].AF
+                        ),
+                        # Decided to use QUALapprox because its formula is easier to interpret than QUAL's
+                        hl.agg.hist(
+                            hl.log10(ht.info.QUALapprox),
+                            *ANNOTATIONS_HISTS["QUALapprox"],
                         ),
-                        hl.agg.hist(hl.log10(ht.info.QUALapprox), 1, 10, 36)
                     )
-                ).map(
-                    lambda x: x[1].annotate(metric=x[0])
-                )
+                ).map(lambda x: x[1].annotate(metric="QUALapprox-"+x[0]))
+            ).extend(
+                hl.array(
+                    hl.agg.group_by(
+                        create_frequency_bins_expr(
+                            AC=ht.freq[1].AC, AF=ht.freq[1].AF
+                        ),
+                        hl.agg.hist(
+                            hl.log10(ht.info.AS_QUALapprox),
+                            *ANNOTATIONS_HISTS["AS_QUALapprox"],
+                        ),
+                    )
+                ).map(lambda x: x[1].annotate(metric="AS_QUALapprox-"+x[0]))
             ),
-            _localize=False
+            _localize=False,
         )
 
-        with hl.hadoop_open(qual_hists_json_path(CURRENT_RELEASE), 'w') as f:
-            f.write(hl.eval(hl.json(hists)))
+        # Defining hist range and bins for allele frequency groups because they needed different ranges
+        inbreeding_bin_ranges = {'under_0.0005': [-0.0005, 0.0005, 50], 'over_0.0005': [-0.25, 1, 50]}
```

Comment:
Could this be parameterized? Or maybe defined as a global after LOG10_ANNOTATIONS so it is a bit easier to read? Or maybe just a new line?

Reply:
I struggled with the best way to handle this because currently the ranges for the other metrics are defined as a json file and each metric has one list with min max and bins. For inbreeding we needed the ranges to be different because of how different the values are compared to frequency. I could either do the global, but then we need to actually modify the code if it changes. Or try changing the code to be able to handle something like this is the json and handle list vs dict differently.

```
{'AS_FS': [0, 30, 50],
 'AS_MQ': [50, 60, 40],
 'AS_MQRankSum': [-5, 5, 60],
 'AS_QD': [0, 40, 40],
 'AS_ReadPosRankSum': [-5, 5, 60],
 'AS_SOR': [0, 5, 50],
 'AS_VarDP': [1, 10, 40],
 'AS_VQSLOD': [-30, 30, 60],
 'AS_pab_max': [0, 1, 50],
 'QUALapprox': [1, 10, 40],
 'AS_QUALapprox': [1, 10, 40],
 'InbreedingCoeff': {"under_0.0005": [-0.0005, 0.0005, 50], "over_0.0005": [-0.25, 1, 50],
}}```

Reply:
@mike-w-wilson thoughts on these options?

Reply:
Agree that sounds good. I was thinking that the other option provided the ability to have multiple metrics with the dict form if needed, and wouldn't require `InbreedingCoeff` to have only under and over, though we have already specified that in `create_frequency_bins_expr_inbreeding` so I actually think this would be better for the current code and more thought could be put in to generalize it later if/when needed

Reply:
Because of some issues in code above this I went with my first suggestion and added a note. I'm still not sure I really like it, but I am struggling to think of better ways. This code is certainly not very generalized, but I suppose it is in the QC repo so its not as big of a deal.

Reply:
That makes sense to me!